### DISCUSSION TITLE
Inference time single sample optimization

### DISCUSSION THREAD

**Discussion by dvirginz** [Lightning-AI/pytorch-lightning#7893] [Lightning Trainer API: Trainer, LightningModule, LightningDataModule]
Can we perform single-sample optimization during testing with unsupervised models using the PL ecosystem?
Given a trained `model`, and a self-supervised `criteria`, I wish to optimize the network for each `test_step` from the end of the training state.
A sketch of the code in pure PyTorch will look something like this:
```
    model = Net()
    for i, data in enumerate(test_dataloader):
        model.load_state_dict(torch.load(PATH))
        while CRITERIA:
            optimizer.zero_grad()
            outputs = model (data)
            loss = self_sup_loss(outputs)
            loss.backward()
            optimizer.step()
```
Should I simply run the inner loop (`while CRITERIA`) inside `test_step` function? 
Thank you.

**Comment by ananthsub** [comment]
@dvirginz - could you describe why this needs to happen during the test step, as opposed to the train step of another lightning module? then you could use manual optimization and handle the while CRITERIA loop inside the train step. https://pytorch-lightning.readthedocs.io/en/latest/common/optimizers.html#manual-optimization

**Reply by dvirginz** [reply]
That's a good idea, correct me if I'm wrong, but no predefined option to load the model weights for each train_step right? I will need to do this manually inside the `train_step`.

**Reply by awaelchli** [reply]
you could keep it in memory and just call `self.load_state_dict(self.initial_weights)` everytime you start a new sample.
